### PR TITLE
RNN

### PR DESCRIPTION
**Pull request by sladenheim**
This PR addresses issue #165. It creates an RNN lecture. The lecture includes

- Description of RNN architecture
- LSTM and GRU architecture
- Example of music generation

### CODE REVIEW

**Comment by sladenheim** [comment]
Updated the code to include the energy prediction example. Feel free to suggest any edits. 

The music file generated is interesting. I'm not sure who to embed the generated music file into the system. I'm also limited on the architecture by my computer and python kernel crashing whenever I try to make the model or batch size too large.

File: ds701_book/27-RNN.qmd [hunk 1]
```diff
@@ -0,0 +1,640 @@
...
+
+However, for longer sequences the influence can be significantly reduced. This is illustrated in red in the figure below.
+
+Due to the potential for long sequences of data, during training you are likely to encounter small gradients. As a result, RNNs are prone to vanishing gradients.
+
+![](drawio/VanishingGradient.png)
+
+## RNN Limitations
+
+- **Vanishing Gradients**:
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:106]: i didn't double check, but is it possible to have exploding gradients too?

File: ds701_book/27-RNN.qmd [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+## Cell State
+
+:::: {.columns}
+::: {.column width="50%"}
+
+The cell state $c_{t-1}$ is the input to the LSTM block.
+
+This value then moves through the LSTM block.
+
+It is modified by either a multiplication or addition interaction.
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:151]: ...and then added to the hidden state after tanh activation.

File: ds701_book/27-RNN.qmd [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+
+:::: {.columns}
+::: {.column width="50%"}
+The forget layer computes a value 
+
+$$
+f_t = \sigma(W_f[h_{t-1}, x_t] + b_f),
+$$
+
+where $\sigma$ is the sigmoid function.
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:171]: Should we add a reminder of what the sigmoid function is?

File: ds701_book/27-RNN.qmd [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+o_t &= \sigma(W_o[h_{t-1}, x_t] + b_o), \\
+\tilde{c}_{t} &=o_t * \operatorname{tanh}(c_t).
+\end{align}
+$$
+
+The vector $o_t$ from the sigmoid layer tells us what parts of the cell state we use for output. The output is a filtered version of the hyperbolic tangent of cell state $c_t$. 
+
+The output of the block is $y_t$. It is the same as the hidden state $h_t$ that is sent to the $t+1$ LSTM block.
+
+This value tells us what *new* information to add to the cell state.
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:231]: "hidden" state, not "cell" state, right?

File: ds701_book/27-RNN.qmd [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:232]: Perhaps list the total trainable weight matrices and bias vectors. 4 of each, right?

File: ds701_book/27-RNN.qmd [hunk 6]
```diff
@@ -0,0 +1,640 @@
...
+- **Long-Term Dependencies**:
+  - RNNs struggle to capture dependencies that are far apart in the sequence, leading to poor performance on tasks requiring long-term memory.
+
+- **Difficulty in Parallelization**:
+  - The sequential processing of data in RNNs makes it challenging to parallelize training, leading to slower training times.
+
+- **Limited Context**:
+  - Standard RNNs have a limited ability to remember context over long sequences, which can affect their performance on complex tasks.
+
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:117]: We should list the bottleneck problem too. The hidden state is all that carries forward history and it can be a bottleneck in how expressive it can be.

File: ds701_book/27-RNN.qmd [hunk 7]
```diff
@@ -0,0 +1,640 @@
...
+test_dataset = TimeSeriesDataset(test_data_scaled, seq_length)
+
+# Create DataLoader
+train_loader = DataLoader(train_dataset, batch_size=8, shuffle=True)
+test_loader = DataLoader(test_dataset, batch_size=8, shuffle=False)
+```
+
+## LSTM Architecture
+
+We define our LSTM architecture in the code cell below.
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:375]: I wonder if there is a graph visualization tool that can show a few of the repeating cells...

File: ds701_book/27-RNN.qmd [hunk 8]
```diff
@@ -0,0 +1,640 @@
...
+plt.figure(figsize=(10, 5))
+plt.plot(range(1, epochs + 1), train_losses, marker='o', linestyle='-', color='b')
+plt.title('Training Loss over Epochs')
+plt.xlabel('Epochs')
+plt.ylabel('Loss')
+plt.xticks(range(1, epochs + 1))  # Set x-tick marks to be integers
+plt.grid(True)
+plt.show()
+```
+
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:430]: We can add a question here to ask the class to make observations on the training curve. Did it converge?

File: ds701_book/27-RNN.qmd [hunk 9]
```diff
@@ -0,0 +1,640 @@
...
+true_rescaled = scaler.inverse_transform(trues)
+
+# Plot results
+plt.figure(figsize=(10, 6))
+plt.plot(true_rescaled, label='True Values')
+plt.plot(predictions_rescaled, label='Predicted Values', alpha=0.7)
+plt.legend()
+plt.show()
+```
+
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:457]: Can we add some interpretations of this plot? Another opportunity for class input. Did it do well? Any obvious problems?

File: ds701_book/27-RNN.qmd [hunk 10]
```diff
@@ -0,0 +1,640 @@
...
+    for instrument in midi_data.instruments:
+        if not instrument.is_drum:
+            for note in instrument.notes:
+                notes.append([note.start, note.end, note.pitch, note.velocity])
+    return np.array(notes)
+```
+
+## Maestro dataset
+
+The [Maestro](https://magenta.tensorflow.org/datasets/maestro#v300) dataset is a dataset composed of about 200 hours of virtuosic piano performances. I took a small sample from one of these files to train our LSTM and generate some new music.
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:489]: Could we add an html control to play examples from the dataset?

File: ds701_book/27-RNN.qmd [hunk 11]
```diff
@@ -0,0 +1,640 @@
...
+all_notes = []
+for midi_file in midi_files:
+  notes = midi_to_notes(midi_file)
+  all_notes.append(notes)
+all_notes = np.concatenate(all_notes)
+```
+
+## Sequences of Notes
+
+Convert the note sequences into a format suitable for training the LSTM model. This means we create sequences of notes and corresponding targets.
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:506]: Can we illustrate this somehow? So midi is describing the notes, kind of like PDF portable document format, and what we want are time based samples, right?

File: ds701_book/27-RNN.qmd [hunk 12]
```diff
@@ -0,0 +1,640 @@
...
+
+X = np.array(X)
+y = np.array(y)
+print("Shape of X", X.shape)
+print("Shape of y", y.shape)
+```
+
+## LSTM Architecture
+
+We now create the LSTM architecture.
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:526]: Can we draw this architecture?

I was going to mention this earlier but in the RNN Theory we should mention that you can stack RNN layers.

File: ds701_book/27-RNN.qmd [hunk 13]
```diff
@@ -0,0 +1,640 @@
...
+
+# Example usage
+start_sequence = X[10]  # Starting sequence for generation
+generated_music = generate_music(model, start_sequence, 100)  # Generate 100 notes
+print(generated_music.shape)
+```
+
+## Download the Song
+
+We can generate an output midi file to play. The song is somewhat similar to the original piece. We could probably create a better song through more training and a larger training set.
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:606]: Again, could we add an html control to play the output? I can help you with that if you want.

File: ds701_book/27-RNN.qmd [hunk 14]
```diff
@@ -0,0 +1,640 @@
...
+#notes_to_midi(generated_music, 'generated_music.midi')
+```
+
+## Recap
+
+We discussed the basic RNN architecture. We then discussed the LSTM and GRU modifications. These modifications allowed the RNNs to handle long-term dependencies.
+
+We then considered two example applications. One of energy consumption prediction and one of of music generation. We generated a piano solo that sounded *interesting*.
+
+One computational bottleneck of RNNs is that they only allow for sequential computation. This was a major bottleneck for applications involving natural language processing. This motivated the development of more advanced language processing architectures, in particular, the transformer, which is the basis for modern large language models (LLMs).
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:634]: We should mention that the architecture evolved to add attention (cite paper), and then famously got rid of the RNN part and just kept attention and the linear layers in "Attention is all you need" and hence the Transformer. I still think RNNs can be useful however.

Comment (trgardos) [reply ds701_book/27-RNN.qmd:634]: Maybe we should allude to his in the intro material too. What do you think?

**Review by trgardos — CHANGES_REQUESTED** [review]
Overall, looks good! Comments embedded.

File: ds701_book/27-RNN.qmd (hunk 10 above)
Comment (trgardos) [reply ds701_book/27-RNN.qmd:489]: added in my PR

File: ds701_book/27-RNN.qmd [hunk 15]
```diff
@@ -0,0 +1,640 @@
...
+        targets = torch.tensor(y[i:i+batch_size], dtype=torch.float32)
+        
+        outputs = model(inputs)
+        loss = criterion(outputs, targets)
+        
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+```
+
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:582]: Plot the training accuracy or loss?

File: ds701_book/27-RNN.qmd (hunk 13 above)
Comment (trgardos) [reply ds701_book/27-RNN.qmd:606]: added in my PR to this branch, but the result is not good.

Comment (trgardos) [reply ds701_book/27-RNN.qmd:606]: Oh, actually, if the model is generating music samples, it would be better to just convert the generated music sample to a wav file and skip MIDI, no?

**Review by trgardos — CHANGES_REQUESTED** [review]

File: ds701_book/27-RNN.qmd (hunk 13 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd:606]: Yes, I would need help with this.

File: ds701_book/27-RNN.qmd (hunk 1 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd:106]: Yes, I believe both are possible though I think that vanishing gradients are more common.

File: ds701_book/27-RNN.qmd (hunk 13 above)
Comment (trgardos) [reply ds701_book/27-RNN.qmd:606]: Ok. can you explain what you are generating? Is midi_to_notes converting to simply time sampled audio data? do you know what sampling frequency?

Comment (sladenheim) [reply ds701_book/27-RNN.qmd:606]: The model is predicting note start, note end, note pitch and note velocity. This function should be taking this generated data and converting it to a midi file. I'm not sure what the sampling frequency is. I used copilot to help write this function and I'm not entirely sure what the prettymidi library is doing under the hood, aside from generating these .midi files. I would need to look this up in more detail.

Comment (trgardos) [reply ds701_book/27-RNN.qmd:606]: Oh. i see from prettyMIDI that it creates numpy arrays of start time, end time, pitch and velocity sequences... Is that what you intended to give as input to the RNN? I had just assumed you were converting regularly sampled audio data.

Comment (sladenheim) [reply ds701_book/27-RNN.qmd:606]: Yes, I'm training the model on that data. I am adding more to the description in my slides.

File: ds701_book/27-RNN.qmd (hunk 14 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd:634]: I was thinking of this as more of setup for the next NLP lectures, but I can mention it in the beginning too.

Comment (sladenheim) [reply ds701_book/27-RNN.qmd:634]: Do you want me to cite the attention is all you need paper? OR is there a different paper you had in mind?

File: ds701_book/27-RNN.qmd (hunk 7 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd:375]: I attempted to draw this using both torchviz and torchview and it looks like this. I'm not sure it's worth showing.
LSTMModel

File: ds701_book/27-RNN.qmd (hunk 14 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd:689]: Is this the paper to cite: 

D. Bahdanau, K. Cho, and Y. Bengio, “Neural Machine Translation by Jointly Learning to Align and Translate.” arXiv, May 19, 2016. Accessed: Feb. 16, 2023. [Online]. Available: http://arxiv.org/abs/1409.0473
2014?

File: ds701_book/27-RNN.qmd (hunk 3 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd:171]: Added.

File: ds701_book/27-RNN.qmd (hunk 5 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd:232]: Added a weight summary slide.

File: ds701_book/27-RNN.qmd (hunk 6 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd:117]: Added.

File: ds701_book/27-RNN.qmd (hunk 8 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd:430]: Added.

File: ds701_book/27-RNN.qmd (hunk 9 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd:457]: Added questions to prompt discussion.

File: ds701_book/27-RNN.qmd (hunk 10 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd:489]: Merged.

File: ds701_book/27-RNN.qmd (hunk 12 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd:562]: See above comment regarding drawing the architecture/compute graph.

File: ds701_book/27-RNN.qmd (hunk 13 above)
Comment (trgardos) [reply ds701_book/27-RNN.qmd:642]: Got it. And I see this Transformer model is using a similar structure of the data too: https://magenta.tensorflow.org/maestro-wave2midi2wave

File: ds701_book/27-RNN.qmd (hunk 7 above)
Comment (trgardos) [reply ds701_book/27-RNN.qmd:375]: yeah, that is too detailed.

File: ds701_book/27-RNN.qmd (hunk 14 above)
Comment (trgardos) [reply ds701_book/27-RNN.qmd:689]: yes.

File: ds701_book/27-RNN.qmd (hunk 15 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd:683]: Added a training loss plot.

File: ds701_book/27-RNN.qmd (hunk 14 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd]: I added a What's Next slide at the end to include this information. I feel that adding all this in the beginning will require too much explanation and the segue into transformers and NLP makes more sense at the end.

File: ds701_book/27-RNN.qmd (hunk 11 above)
Comment (sladenheim) [reply ds701_book/27-RNN.qmd]: Added an illustration to show how to break up the data into X and y.

**Comment by sladenheim** [comment]
I removed the content about music generation.

File: ds701_book/27-RNN.qmd [hunk 16]
```diff
@@ -0,0 +1,823 @@
...
+:::: {.columns}
+::: {.column width="50%"}
+Using the a 3-layer RNN with 512 hidded nodes on each layer, Karpathy trained a model on the complete works of Shakespeare. With this model he was able to generate texts similar to what is seen on the left.
+
+Observe that there are some typos and the syntax is not perfect, but overall the style appears Shakespearean.
+
+:::
+::: {.column width="5-%"}
+![](figs/ShakespeareText.png)
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:763]: Figure not found.

File: ds701_book/27-RNN.qmd [hunk 17]
```diff
@@ -0,0 +1,823 @@
...
+
+:::: {.columns}
+::: {.column width="30%"}
+Using a 100MB [dataset](http://prize.hutter1.net) of raw Wikipedia data, Karpathy trained RNNs which generated Wikipedia content. An example of the generated markdown is shown to the right.
+
+The text is nonsense, but is structured like a Wikipedia post.
+Interestingly, the model *hallucinated* and fabricated a url that does not exist.
+:::
+::: {.column width="70%"}
+![](figs/WikipediaContent.png)
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:777]: Missing figure

File: ds701_book/27-RNN.qmd [hunk 18]
```diff
@@ -0,0 +1,823 @@
...
+::: {.column width="30%"}
+Using a list of 8000 baby names from this [link](http://www.cs.cmu.edu/afs/cs/project/ai-repository/ai/areas/nlp/corpora/names/), Karpathy trained an RNN to predict baby names.
+
+To the right is a list of generated names not on the lists. More of the names can be found [here](https://cs.stanford.edu/people/karpathy/namesGenUnique.txt).
+
+Some interesting additional names the model generated were
+`Baby, Char, Mars, Hi, and With`.
+:::
+::: {.column width="70%"}
+![](figs/BabyNames.png)
```
Comment (trgardos) [inline ds701_book/27-RNN.qmd:794]: Missing figure.

**Review by trgardos — CHANGES_REQUESTED** [review]
please add missing figures

File: ds701_book/27-RNN.qmd (hunk 16 above)
Comment (trgardos) [reply ds701_book/27-RNN.qmd:763]: never mind. an issue on my side.

File: ds701_book/27-RNN.qmd (hunk 17 above)
Comment (trgardos) [reply ds701_book/27-RNN.qmd:777]: issue on my side. disregard

File: ds701_book/27-RNN.qmd (hunk 18 above)
Comment (trgardos) [reply ds701_book/27-RNN.qmd:794]: disregard

**Review by trgardos — APPROVED** [review]
Looks good!
